{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accounting human miRNA overhangs\n",
    "\n",
    "Input file `hsa.gff3` obtained from miRBase: \n",
    "<https://www.mirbase.org/ftp/CURRENT/genomes/hsa.gff3>\n",
    "\n",
    "## Creating individual files \n",
    "\n",
    "First of all, different files for each feature - pre-miRs and mature miRNAs - and split the last one into 3 different files:\n",
    "- `mirna_5p.gff3` for the 5p mature miRNAs. \n",
    "- `mirna_3p.gff3` for the 3p mature miRNAs.  \n",
    "- `mirna_single.gff3` for the pre-miRs with a single mature miRNA.\n",
    "\n",
    "In addition, names in column 9 will be formatted to contain only the name - no ID, no alias."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " There are:\n",
      "- 1918 miRNA primary transcripts\n",
      "- 992 mature miRNAs annotated as 5p\n",
      "- 992 mature miRNA annotated as 3p\n",
      "- 899 mature miRNA annotated to be the only transcript of the pre-miRNA\n"
     ]
    }
   ],
   "source": [
    "## Reading file\n",
    "with open(\"hsa.gff3\", \"r\") as f:\n",
    "    records = [line.strip().split() for line in f.readlines() if line[0] != \"#\"]\n",
    "\n",
    "## Formating last col to contain only the name\n",
    "for rec in records:\n",
    "    rec[-1] = rec[-1].split(\"=\")[3].split(\";\")[0]\n",
    "\n",
    "## Nesting lists to create files\n",
    "pre_mirna = [rec for rec in records if rec[2] == \"miRNA_primary_transcript\"]\n",
    "mirna_5p = [rec for rec in records if rec[2] == \"miRNA\" and rec[-1].split(\"-\")[-1] == \"5p\"]\n",
    "mirna_3p = [rec for rec in records if rec[2] == \"miRNA\" and rec[-1].split(\"-\")[-1] == \"3p\"]\n",
    "mirna_single = [rec for rec in records if rec[2] == \"miRNA\" and rec[-1].split(\"-\")[-1] != \"5p\" and rec[-1].split(\"-\")[-1] != \"3p\"]\n",
    "\n",
    "## Removing transcript number from the name     \n",
    "for record in pre_mirna:\n",
    "    name = record[-1].split(\"-\")\n",
    "    record[-1] = \"-\".join(name[:3])\n",
    "\n",
    "print(f\" There are:\\n- {len(pre_mirna)} miRNA primary transcripts\")\n",
    "print(f\"- {len(mirna_5p)} mature miRNAs annotated as 5p\")\n",
    "print(f\"- {len(mirna_3p)} mature miRNA annotated as 3p\")\n",
    "print(f\"- {len(mirna_single)} mature miRNA annotated to be the only transcript of the pre-miRNA\")\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the [Accounting human miRNA annotation file](https://github.com/zavolanlab/mirflowz/blob/mirna_accounting/explorations/accounting_mirna_annotations.ipynb) there are the same amount of mature miRNA records, 2883, but the distribution among `5p`, `3p` and `single` are different.\n",
    "There are 27 more entries in both the `5p` and the `3p` therefore 54 less entries counted as single mature miRNAs.\n",
    "\n",
    "This difference takes place due to some single-mature miRNA being annotated as `5p` or `3p`. \n",
    "If subtracting the extra sequences and adding them where they belong, the they sum up to the number of pre-mirs, 1918 "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Solution\n",
    "\n",
    "Given this situation, the filtering must be done in another way. We know that if a pre-miR has two mature sequences, the features in the `gff3` file will be ordered as:\n",
    "\n",
    "```\n",
    "miRNA_primary_transcript_1\n",
    "miRNA_1\n",
    "miRNA_1\n",
    "miRNA_primary_transcript_2\n",
    "```\n",
    "\n",
    "On the other hand, if the pre-miRNA has a single mature form, the `gff3` will look:\n",
    "\n",
    "```\n",
    "miRNA_primary_transcript_1\n",
    "miRNA_1\n",
    "miRNA_primary_trasncript_2\n",
    "```\n",
    "This configuration helps creating the files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " There are:\n",
      "- 1918 miRNA primary transcripts\n",
      "- 965 mature miRNAs annotated as 5p\n",
      "- 965 mature miRNA annotated as 3p\n",
      "- 953 mature miRNA annotated to be the only transcript of the pre-miRNA\n"
     ]
    }
   ],
   "source": [
    "## Creating empty lists to store entries\n",
    "pre_mirna = []\n",
    "mirna_5p = []\n",
    "mirna_3p = []\n",
    "mirna_single = []\n",
    "\n",
    "## Setting values for previous and next entries\n",
    "prev_rec = None\n",
    "pos_rec = None\n",
    "\n",
    "## For each record except the last two...\n",
    "for idx in range(len(records) - 2):\n",
    "\n",
    "    ## ..if first entry..\n",
    "    if prev_rec == None:\n",
    "        prev_rec = records[idx][2]\n",
    "        pos_rec = records[idx + 2][2]\n",
    "        pre_mirna.append(records[idx])\n",
    "\n",
    "    ## ..otherwise..\n",
    "    else:\n",
    "        if records[idx][2] == \"miRNA_primary_transcript\":\n",
    "            pre_mirna.append(records[idx])\n",
    "        else:\n",
    "            if prev_rec == pos_rec:\n",
    "                mirna_single.append(records[idx])\n",
    "            else:\n",
    "                if records[idx][-1].split(\"-\")[-1] == \"5p\":\n",
    "                    mirna_5p.append(records[idx])\n",
    "                else:\n",
    "                    mirna_3p.append(records[idx])\n",
    "        prev_rec = records[idx][2]\n",
    "        pos_rec = records[idx + 2][2]\n",
    "\n",
    "## Reading last two entries\n",
    "if records[-2][2] == \"miRNA_primary_transcript\":\n",
    "    pre_mirna.append(records[-2])\n",
    "    mirna_single.append(records[-1])\n",
    "else:\n",
    "    if records[-2][-1].split(\"-\")[-1] == \"5p\":\n",
    "        mirna_5p.append(records[-2])\n",
    "        mirna_3p.append(records[-1])\n",
    "    else:\n",
    "        mirna_3p.append(records[-2])\n",
    "        mirna_5p.append(records[-1])\n",
    "\n",
    "\n",
    "print(f\" There are:\\n- {len(pre_mirna)} miRNA primary transcripts\")\n",
    "print(f\"- {len(mirna_5p)} mature miRNAs annotated as 5p\")\n",
    "print(f\"- {len(mirna_3p)} mature miRNA annotated as 3p\")\n",
    "print(f\"- {len(mirna_single)} mature miRNA annotated to be the only transcript of the pre-miRNA\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we got the correct classification counts, let's count the nucleotides' overhang per sequence in the form of dictionaries.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#### REFORMATTING ENTRIES TO EASE THE COUNT ####\n",
    "\n",
    "end3_clean = []\n",
    "end5_clean = []\n",
    "single_clean = []\n",
    "pre_clean = []\n",
    "\n",
    "for entry in mirna_3p:\n",
    "    name = entry[-1]\n",
    "    name = name.split(\"-\")[:3]\n",
    "    name = \"-\".join(name)\n",
    "    if entry[6] == '-':\n",
    "        end3_clean.append([name.lower(), entry[3], entry[0], entry[6]])\n",
    "    else:\n",
    "        end3_clean.append([name.lower(), entry[4], entry[0], entry[6]])\n",
    "\n",
    "for entry in mirna_5p:\n",
    "    name = entry[-1]\n",
    "    name = name.split(\"-\")[:3]\n",
    "    name = \"-\".join(name)\n",
    "    if entry[6] == '-':\n",
    "        end5_clean.append([name.lower(), entry[4], entry[0], entry[6]])\n",
    "    else:\n",
    "        end5_clean.append([name.lower(), entry[3], entry[0], entry[6]])\n",
    "\n",
    "for entry in mirna_single:\n",
    "    name = entry[-1]\n",
    "    name = name.split(\"-\")[:3]\n",
    "    name = \"-\".join(name)\n",
    "    single_clean.append([name.lower(), entry[3], entry[4], entry[0], entry[6]])\n",
    "\n",
    "for entry in pre_mirna:\n",
    "    name = entry[-1]\n",
    "    name = name.split(\"-\")[:3]\n",
    "    name = \"-\".join(name)\n",
    "    pre_clean.append([name, entry[3], entry[4], entry[0], entry[6]])\n",
    "\n",
    "\n",
    "#### MAIN PART: CREATING TABLES ####\n",
    "\n",
    "count_5 = {}\n",
    "count_3 = {}\n",
    "total_5 = 0\n",
    "total_3 = 0\n",
    "\n",
    "\n",
    "for mir in pre_clean:\n",
    "    for miR3 in end3_clean:\n",
    "        if mir[0] == miR3[0] and mir[1] <= miR3[1] <= mir[2] and mir[3] == miR3[2]:\n",
    "            total_3 += 1\n",
    "            ntl = str(int(mir[2]) - int(miR3[1]))\n",
    "            if int(ntl) > 20:\n",
    "            \tntl = \"20+\"\n",
    "            if ntl in count_3.keys():\n",
    "                count_3[ntl] += 1\n",
    "            else:\n",
    "                count_3[ntl] = 1\n",
    "            idx = end3_clean.index(miR3)\n",
    "            del end3_clean[idx]\n",
    "            break\n",
    "\n",
    "    for miR5 in end5_clean:\n",
    "        if mir[0] == miR5[0] and mir[1] <= miR5[1] <= mir[2] and mir[3] == miR5[2]:\n",
    "            total_5 += 1\n",
    "            ntl = str(int(miR5[1]) - int(mir[1]))\n",
    "            if int(ntl) > 20:\n",
    "            \tntl = \"20+\"\n",
    "            if ntl in count_5.keys():\n",
    "                count_5[ntl] += 1\n",
    "            else:\n",
    "                count_5[ntl] = 1\n",
    "            idx = end5_clean.index(miR5)\n",
    "            del end5_clean[idx]\n",
    "            break\n",
    "    \n",
    "    for miR in single_clean:\n",
    "        if mir[0] == miR[0] and mir[1] <= miR[1] and mir[2] >= miR[2] and mir[3] == miR[3]:\n",
    "            total_5 += 1\n",
    "            total_3 += 1\n",
    "            ntl_3 = str(int(mir[2]) - int(miR[2]))\n",
    "            ntl_5 = str(int(miR[1]) - int(mir[1]))\n",
    "            if int(ntl_3) > 20:\n",
    "            \tntl_3 = \"20+\"\n",
    "            if int(ntl_5) > 20:\n",
    "            \tntl_5 = \"20+\"\n",
    "            if ntl_5 in count_5.keys():\n",
    "                count_5[ntl_5] += 1\n",
    "            else:\n",
    "                count_5[ntl_5] = 1\n",
    "            if ntl_3 in count_3.keys():\n",
    "                count_3[ntl_3] += 1\n",
    "            else:\n",
    "                count_3[ntl_3] = 1\n",
    "            idx = single_clean.index(miR)\n",
    "            del single_clean[idx]\n",
    "            break\n",
    "\n",
    "\n",
    "#### WRITTING DATARFAME TO MANIPULATE ON R ####\n",
    "\n",
    "final3 = dict(sorted(count_3.items()))\n",
    "final5 = dict(sorted(count_5.items()))\n",
    "\n",
    "df3 = pd.DataFrame(final3, index=[0])\n",
    "df3.to_csv(\"count_3.csv\", index = False, sep = \"\\t\")\n",
    "\n",
    "df5 = pd.DataFrame(final5, index=[0])\n",
    "df5.to_csv(\"count_5.csv\", index = False, sep = \"\\t\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "What follows is the script used in R to create the tables:\n",
    "\n",
    "```R\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "\n",
    "## Creating table for the 3p overhangs\n",
    "count_3 <- read.csv(\"count_3.csv\", header = F, sep = \"\\t\")\n",
    "count_3 <- as.data.frame(t(count_3))\n",
    "row.names(count_3) <- NULL\n",
    "colnames(count_3) <- c(\"Overhang\", \"Count\")\n",
    "\n",
    "count_3 <- count_3 %>% \n",
    "                mutate(\"Fraction\" = round(as.numeric(Count) /1918, 4)) %>%\n",
    "                arrange(as.numeric(Overhang))\n",
    "count_3$Overhang <- as.numeric(count_3$Overhang)\n",
    "cum3 <- cumsum(count_3$Fraction)\n",
    "count_3 <- count_3 %>%\n",
    "  mutate(\"Comulative Proportion\" = cum3)\n",
    "\n",
    "\n",
    "## Creating table for the 5p overhangs\n",
    "count_5 <- read.csv(\"count_5.csv\", header = F, sep = \"\\t\")\n",
    "count_5 <- as.data.frame(t(count_5))\n",
    "row.names(count_5) <- NULL\n",
    "colnames(count_5) <- c(\"Overhang\", \"Count\")\n",
    "\n",
    "count_5 <- count_5 %>% \n",
    "                mutate(\"Fraction\" = round(as.numeric(Count) /1918, 4)) %>%\n",
    "                arrange(as.numeric(Overhang))\n",
    "count_5$Overhang <- as.numeric(count_5$Overhang)\n",
    "cum5 <- cumsum(count_5$Fraction)\n",
    "count_5 <- count_5 %>%\n",
    "  mutate(\"Comulative Proportion\" = cum5)\n",
    "\n",
    "\n",
    "# Merging tables\n",
    "\n",
    "both <- inner_join(count_5, count_3,by = \"Overhang\", suffix = c(\"_5p\", \"_3p\")) %>%\n",
    "        arrange(as.numeric(Overhang))\n",
    "  \n",
    "both$Overhang[22] <- \"20+\"\n",
    "\n",
    "\n",
    "\n",
    "write.table(both, \"overhang.txt\", quote = F, sep = \"\\t\", row.names = F)\n",
    "```"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final table looks like:\n",
    "\n",
    "|Overhang (nts) |Count_5p | Proportion_5p| Cumulative_5p|Count_3p | Proportion_3p| Cumulative_3p| \n",
    "|:--------------|:--------|-------------:|-------------:|:--------|-------------:|-------------:| \n",
    "|0              |83       |        0.0433|        0.0433|139      |        0.0725|        0.0725| \n",
    "|1              |19       |        0.0099|        0.0532|28       |        0.0146|        0.0871| \n",
    "|2              |27       |        0.0141|        0.0673|37       |        0.0193|        0.1064| \n",
    "|3              |25       |        0.0130|        0.0803|34       |        0.0177|        0.1241| \n",
    "|4              |33       |        0.0172|        0.0975|37       |        0.0193|        0.1434| \n",
    "|5              |102      |        0.0532|        0.1507|24       |        0.0125|        0.1559| \n",
    "|6              |21       |        0.0109|        0.1616|32       |        0.0167|        0.1726| \n",
    "|7              |31       |        0.0162|        0.1778|35       |        0.0182|        0.1908| \n",
    "|8              |36       |        0.0188|        0.1966|39       |        0.0203|        0.2111| \n",
    "|9              |95       |        0.0495|        0.2461|63       |        0.0328|        0.2439| \n",
    "|10             |111      |        0.0579|        0.3040|133      |        0.0693|        0.3132| \n",
    "|11             |26       |        0.0136|        0.3176|51       |        0.0266|        0.3398| \n",
    "|12             |40       |        0.0209|        0.3385|53       |        0.0276|        0.3674| \n",
    "|13             |35       |        0.0182|        0.3567|33       |        0.0172|        0.3846| \n",
    "|14             |49       |        0.0255|        0.3822|33       |        0.0172|        0.4018| \n",
    "|15             |88       |        0.0459|        0.4281|63       |        0.0328|        0.4346| \n",
    "|16             |14       |        0.0073|        0.4354|7        |        0.0036|        0.4382| \n",
    "|17             |8        |        0.0042|        0.4396|17       |        0.0089|        0.4471| \n",
    "|18             |7        |        0.0036|        0.4432|10       |        0.0052|        0.4523| \n",
    "|19             |12       |        0.0063|        0.4495|14       |        0.0073|        0.4596| \n",
    "|20             |17       |        0.0089|        0.4584|15       |        0.0078|        0.4674| \n",
    "|20+            |1036     |        0.5401|        0.9985|1021     |        0.5323|        0.9997|"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
